mypy/state.py

```diff
@@ -24,6 +27,15 @@ def strict_optional_set(self, value: bool) -> Iterator[None]:
         finally:
             self.strict_optional = saved
 
+    @contextmanager
+    def type_checker_set(self, value: TypeCheckerSharedApi) -> Iterator[None]:
```

Comment:
Dependency on `TypeCheckerSharedApi` probably makes various import cycles worse, and I assume this why there are some additional nested imports. Defining `type_checker_set` in a new module would improve things, right? Splitting this module seems better than making import cycles bigger, and it should also reduce the performance regression.

Reply:
> Defining `type_checker_set` in a new module would improve things, right? 

Yeah, I think this should be better. I will play with this.